File: admin/src/shared/api/billing.ts
import type { Plan } from '@/shared/api/auth';
import { request } from '@/shared/api/http';

export const createCheckout = (
	plan: Plan,
): Promise<{ checkout_url: string }> =>
	request<{ checkout_url: string }>('/api/v1/billing/checkout', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ plan }),
	});

export const getPortalUrl = (): Promise<{ portal_url: string }> =>
	request<{ portal_url: string }>('/api/v1/billing/portal');

File: backend/app/routes/billing.py
from fastapi import APIRouter, Depends, HTTPException

from app.models.billing import CheckoutRequest, CheckoutResponse, PortalResponse
from app.models.user import UserInDB
from app.routes.deps import get_billing_service, require_user
from app.services.billing_service import BillingService

router = APIRouter(prefix="/billing", tags=["billing"])


@router.post("/checkout", response_model=CheckoutResponse)
async def create_checkout(
    body: CheckoutRequest,
    user: UserInDB = Depends(require_user),
    svc: BillingService = Depends(get_billing_service),
) -> CheckoutResponse:
    return CheckoutResponse(checkout_url=await svc.create_checkout(user, body.plan))


@router.get("/portal", response_model=PortalResponse)
async def portal(
    user: UserInDB = Depends(require_user),
    svc: BillingService = Depends(get_billing_service),
) -> PortalResponse:
    if not user.polar_customer_id:
        raise HTTPException(status_code=404, detail="No subscription on file")
    return PortalResponse(portal_url=await svc.create_portal(user))
